extensions/amp-web-push/0.1/web-push-service.js

```diff
@@ -0,0 +1,697 @@
+   * extension in the example sandbox. Turns off in unit test mode.
+   */
+  enableAmpExperimentForDevelopment_() {
+    if ((getMode().localDev && !getMode().test)) {
+      AMP.toggleExperiment(TAG, true);
+    }
+  }
+
+  log_(text) {
+    // Only prints if the user turned on debug/verbose mode
+    if (this.debug_) {
+      // For easy debugging, print out the origin this code is running on
+      console.log.apply(null, [`${location.origin}: ${text}`]);
+    }
+  }
+
+  /*
+   * Checks that the user enabled this AMP experiment and allows integration
+   * tests to access this class in testing mode.
+   */
+  ensureAmpExperimentEnabled_() {
+    // Allow integration tests to access this class in testing mode.
+    /** @const @private {boolean} */
+    const isExperimentEnabled = isExperimentOn(this.ampdoc.win, TAG);
+    user().assert(isExperimentEnabled, `Experiment "${TAG}" is disabled. ` +
+      `Enable it on ${urls.cdn}/experiments.html.`);
+  }
+
+  /*
+   * Searches for a <script type="application/json"> configuration and returns
+   * it as text.
+   */
+  getConfigAsText() {
+    const configJsonNode =
+      this.ampdoc.getRootNode().querySelector('script#' + TAG);
+    if (!configJsonNode) {
+      throw user().createError('Your AMP document must include a ' +
+        '<script id="amp-web-push" type="application/json">.');
+    }
+    return configJsonNode.textContent;
+  }
+
+  /*
+  * Parses the JSON configuration and returns a JavaScript object. Also
+  * validates the input.
+  */
+  parseConfigJson(configJson) {
+    if (!configJson || !configJson.trim()) {
+      throw user().createError('Your AMP document\'s configuration ' +
+        'JSON must not be empty.');
+    }
+    let config;
+    try {
+      config = parseJson(/** @type {string} */(configJson));
+    } catch (e) {
+      throw user().createError('Your AMP document\'s configuration ' +
+        'JSON must be valid JSON. Failed to parse JSON: ' + e);
+    }
+
+    if (!config['helperIframeUrl'] ||
+      !this.isValidHelperOrPermissionDialogUrl_(
+          config['helperIframeUrl'])) {
+      throw user().createError('Your AMP document\'s configuration JSON ' +
+        'must have a valid helperIframeUrl property. It should begin with ' +
+        'the https:// protocol and point to the provided lightweight ' +
+        'template page provided for AMP messaging.');
+    }
+
+    if (!config['permissionDialogUrl'] ||
+      !this.isValidHelperOrPermissionDialogUrl_(
+          config['permissionDialogUrl'])) {
+      throw user().createError('Your AMP document\'s configuration JSON must ' +
+        'have a valid permissionDialogUrl property. It should begin with ' +
+        'the https:// protocol and point to the provided template page ' +
+        'for showing the permission prompt.');
+    }
+
+    if (!config['serviceWorkerUrl'] ||
+      new URL(config['serviceWorkerUrl']).protocol !== 'https:') {
```

Comment:
Please use `url.js` `parseUrl`. Still enough user agents not supporting URL API.